pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,454 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud"
+	aliclient "github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud/client"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	ctrlerror "github.com/gardener/gardener/pkg/controllerutils/reconciler"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// bastionEndpoints holds the endpoints the bastion host provides
+type bastionEndpoints struct {
+	// private is the private endpoint of the bastion. It is required when opening a port on the worker node to allow SSH access from the bastion
+	private *corev1.LoadBalancerIngress
+	//  public is the public endpoint where the enduser connects to establish the SSH connection.
+	public *corev1.LoadBalancerIngress
+}
+
+// Ready returns true if both public and private interfaces each have either
+// an IP or a hostname or both.
+func (be *bastionEndpoints) Ready() bool {
+	return be != nil && IngressReady(be.private) && IngressReady(be.public)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, a.client, &opt.SecretReference)
+	if err != nil {
+		return err
+	}
+
+	aliCloudECSClient, err := a.newClientFactory.NewECSClient(opt.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+
+	aliCloudVPCClient, err := a.newClientFactory.NewVPCClient(opt.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+
+	imageID, err := getImageID(cluster, opt)
+	if err != nil {
+		return err
+	}
+
+	instanceType, err := aliCloudECSClient.GetInstanceType(cores)
+	if err != nil {
+		return err
+	}
+
+	if instanceType == nil ||
+		len(instanceType.AvailableZones.AvailableZone) == 0 ||
+		len(instanceType.AvailableZones.AvailableZone[0].AvailableResources.AvailableResource) == 0 ||
+		len(instanceType.AvailableZones.AvailableZone[0].AvailableResources.AvailableResource[0].SupportedResources.SupportedResource) == 0 ||
+		instanceType.AvailableZones.AvailableZone[0].AvailableResources.AvailableResource[0].SupportedResources.SupportedResource[0].Status != "Available" {
+		return errors.New("one cores instance is empty or unavailable")
```

Comment:
Currently it is hard-coded that we only use 1-core instance. However, should there really be circumstances where we do not have 1-core instance it would be pain in the back that we cannot create a bastion. Ideally we should be able to manually choose an instance type, this would provide additional option if the default one here doesn't work.

Reply:
agree. Same as for the bastion openstack implementation this should be an extension-provider configuration

Reply:
committed in e4ea144c. add default instance type. it seems maintains in wdf landscape-canary-garden under provider-alicloud resources.

Reply:
We could have a default solution here as currently implemented (we try to get the instance type with least core count). My point is that perhaps we can add a `InstanceType` configuration in Bastion configuration? And for `ssh` command of gardenctl, we get to specify something like `gardenctl_v2 ssh <shoot_name> --bastionInstanceType <instance_type>`. For AliCloud specific command, we could also provide an option to list available instance types.

As for setting it directly in extension-provider configuration, with AliCloud, I think it is not a good solution, since different user in different AZ may have different available instance types. Even the most general ones like `ecs.g6.*` may run out of stock.

Overall this PR looks fine to me and we can start testing it. We could have a separate enhancement PR to resolve above mentioned problem as it also requires changes on gardenctl side.

Reply:
If it's dynamic, I'd rather read it from the cloudprofile instead of having provider specific logic in gardenctl

By the way, on the openstack implementation we have a configuration option (imageRef and flavorRef) https://github.com/gardener/gardener-extension-provider-openstack/blob/a281d28ee0f45b7bae0a1e5e31e067d474a6c9c0/example/00-componentconfig.yaml#L17-L19

Reply:
some comments here, as the latest change in https://github.com/gardener/gardener-extension-provider-alicloud/commit/c13fde31c2d9bc0332aa58ff785da7f6efe8d567 . the new logic will searching the lowest amount of CPUs from an available zone. it has about ~20 different instance types from cores <=2 when I testing depends on the available zone. It might mitigate the issue